vundo-diff

jdtsmith opened the pull request:
This provides `vundo-diff` support with the following capabilities:

- ~It's optional, enabled by the `vundo-diff-mode` global minor mode~ (now builtin)
- Adds keys `(m)ark`, `(u)nmark`, and `(d)iff`
- Marked node has a separate (new) color, blue by default
- Diff pair defaults to parent if no node is marked
- Automatically swaps states based on index
- Includes additional color-coded header information in the diff file, including save timestamp (if available)

casouri commented on vundo-diff.el:
```diff
@@ -0,0 +1,157 @@
+	(display-buffer dbuf)))))
+
+;;;###autoload
+(define-minor-mode vundo-diff-mode
```
That way we don't need a minor mode.

casouri commented:
Thanks, looks nice! But please make sure only indent with spaces, I'm seeing tabs again. Apart from the comments above, I also see terminal color codes in the terminal. I know you are using Emacs facility for diffing, is there anything we can do? Or at least teach users how to resolve it in the docs.

<img width="629" alt="Screenshot 2023-12-03 at 11 27 43 PM" src="https://github.com/casouri/vundo/assets/13118647/0ba3dc6f-4073-4b4d-ae09-a383e35e67e6">

jdtsmith commented:
Thanks. Indent fixed.  For your color codes in terminal, maybe try `(setq diff-switches "-u --color=never")`? 

The one other feature I considered overnight is just replacing those terrible tmp file names with the Current/Marked (with TS).  But it sounds a bit fragile, so maybe not worth it for the improved cleanliness.

jdtsmith commented on vundo-diff.el:
```diff
@@ -0,0 +1,150 @@
+        (message "vundo diff not available.")
+      (let ((mrkbuf (get-buffer-create
+                     (make-temp-name (concat oname "-vundo-diff-marked")))))
+        (vundo--move-to-node current marked orig vundo--prev-mod-list)
+        (with-current-buffer mrkbuf
+          (insert-buffer-substring-no-properties orig))
+        (vundo--refresh-buffer orig (current-buffer) 'incremental)
+        (vundo--move-to-node marked current orig vundo--prev-mod-list)
+        (vundo--refresh-buffer orig (current-buffer) 'incremental)
```
I want to double-check.  Is it your understanding that you need to `(vundo--refresh-buffer ... 'incremental)` after every `move-to-node` like this?   Somehow I got the impression that this is needed only once (i.e. you can move as much as you like).  But I get `aref` and `no route` errors without both of these calls.

jdtsmith replied:
I take it back.  I get `no route` without a `refresh-buffer` after each move.  Restored that.  Any idea why?

jdtsmith commented:
I have added a sanity check wrapper prior to the moves, and pushed some README/NEWS changes for your consideration.  I'm still playing with the formatting in the diff buffer, will wrap that up soon.

jdtsmith commented:
I made a few more improvements today:

- Turns out that diff can be called async, and if so, I need to update/cleanup the diff buffer in the process sentinel.
- Since `vundo-diff` displays timestamps independent of whether the user has green circles on (`vundo-highlight-saved-nodes`), we need `last-saved-idx` to always be saved (further separating display/other logic).  
- Similarly, properly report the `last-saved-idx` timestamp for the node, when it is just the most recent saved buffer state — it gets a green circle, after all.  (New function `vundo--any-timestamp`). 
- Cleaned up the `vundo-diff` results buffer to remove the ugly tmp file names and substitute vundo-specific info. (New function `vundo-diff--cleanup-diff-buffer`).
- Created a tiny `vundo-diff-mode` derived from `diff-mode`, to fontify said file header info.
- Added `declare-function`'s to silence compiler.  

This PR is pretty well complete IMO, pending any further review.

casouri commented on vundo.el:
```diff
@@ -513,6 +513,41 @@ If FROM non-nil, build from FORM-th modification in MOD-LIST."
+                     (aref mod-list next-mod-idx))))
+    (and next-mod (vundo-m-timestamp next-mod))))
+
+(defun vundo--any-timestamp (mod-list node)
```
Maybe this can be `vundo--node-timestamp`?

casouri commented:
Looks nice!

> Turns out that diff can be called async, and if so, I need to update/cleanup the diff buffer in the process sentinel.

Why use the async version? IMO the sync version is perfectly fine and should be simpler to use, right?

And I wonder if we can use a single key to mark/unmark nodes, instead of using both `m` and `u`. You know, make it `vundo-toggle-mark` or something.

jdtsmith commented:
> > Turns out that diff can be called async, and if so, I need to update/cleanup the diff buffer in the process sentinel.
> 
> Why use the async version? IMO the sync version is perfectly fine and should be simpler to use, right?

Since it's the default, people might be used to that behavior of diff.  The cost to support async is not high: just a lambda for the process sentinel.  And sometimes tmp files are across a tramp link.  

> And I wonder if we can use a single key to mark/unmark nodes, instead of using both `m` and `u`. You know, make it `vundo-toggle-mark` or something.

Good idea.  Maybe just `m`.

And if you don't mind taking a look up at my question about `vundo-refresh-buffer`, I'm still a bit confused about the need for that to be there twice.

casouri commented:
> We could make it so that m on an unmarked node sets is, m on a marked node unsets. But that's getting fancy/non-intuitive. I think m + u is pretty simple and clean, and leverages dired knowledge.

Ok, I trust your judgement here.

> And if you don't mind taking a look up at my question about vundo-refresh-buffer, I'm still a bit confused about the need for that to be there twice.

Yeah, moving to a node pushes new undo entries onto undo-list, and we need to process those with `vundo-refresh-buffer` (update mod-list, etc).

jdtsmith commented:
Thanks for the feedback.

> > And if you don't mind taking a look up at my question about vundo-refresh-buffer, I'm still a bit confused about the need for that to be there twice.
> 
> Yeah, moving to a node pushes new undo entries onto undo-list, and we need to process those with `vundo-refresh-buffer` (update mod-list, etc).

I think what confuses me is this note from the `vundo--trim-undo-list` doc:

> We can call `vundo--move-to-node' multiple times between two `vundo--refresh-buffer'. 

That doesn't seem to be the case here.  I.e. you must call `refresh-buffer` after each of the pair of moves (there and back) or you get `no route`.

Found one more little doc tweak I pushed, so this is ready to go.   

In terms of version bump, I do have a couple of small additional fixes I'd like to send in as tiny PRs once this is merged:

1. Make `l` key move to a _prior_ saved node, if already on one (so you can `l`, `l`, `l`, to go back through saved states).
2. Ensure the pattern _undo-to-node, save it, alter it_ correctly marks that node as _saved_: this requires checking the equivalents of the node for any subsequent TS-ful mod.  This does not solve the issue outlined in #66, but that is rare in practice.

casouri commented:
> That doesn't seem to be the case here. I.e. you must call refresh-buffer after each of the pair of moves (there and back) or you get no route.

Oh, I would trust the me that wrote the comment much more than I trust the me right now. If the comment says so, we should be able to run `move-to-node` multiple times.

I think I know why you'll get "no route". One example: in an undo list [1 2 3 4 5], you'd have mod list [1 2 3 4 5], and if you go back to node 3, your undo list becomes [1 2 3 4 5 4' 3'], but without calling `refresh`, the mod list remains [1 2 3 4 5]. Now, if you want to go back to 5, there's no route available in the mod list (you can think of it as you can only go left in the list, so from 5 to 1-4, but not the other way around). Once you call `refresh`, the mod list becomes [1 2 3 4 5 4' 3'], now you have a route from 3 to 5 (in the form of 3' to 5, which is going left).

Good call though, I never thought about this before. I should add a comment somewhere calling this out.

So the branch is ready for a final review, right? It'd be great if you can squash everything and write a commit message. Then I'll review and merge it.

jdtsmith commented:
> I think I know why you'll get "no route". 

Right, you can't change direction between `refresh`'s.

> I'll review and merge it.

Great thanks, commits squashed, ready for review.   Once merged I'll submit my small PR for timestamp stuff (I'm already loving multiple `l`).

casouri commented:
Ok, some final nits:
- Could you leave the defvar of `vundo--orig-buffer` and `vundo--last-saved-idx` in their original place, and just place `(defvar vundo--last-saved-idx)` in the timestamp section (to suppress compiler errors)?
- Could you write the commit message in GNU format, such that it explain what functions/variable are changed? For the new file (vundo-diff.el), simply saying "new file" is enough.
- I feel like the "current node" highlight (red bold) should override any other highlights, including the "marked node" highlight.
- Could you put `(setq diff-switches "-u --color=never")` somewhere in the documentation?

jdtsmith commented:
OK should be all set, thanks for the review.  I also threw in a docstring escape of single quotes to silence the compiler.  Once this is merged I have a TS-related PR I'll post.

casouri commented:
Merged, thanks!
